### ISSUE TITLE
Different results on Spyder vs Visual Studio

### ISSUE THREAD

**Issue by rummeln** [StatguyUser/BaselineRemoval#15]
Hi, I've been trying to troubleshoot this for a few days, but can't figure it out.  I'm running the same code using baselineremoval with imodpoly, but I'm getting different results on Visual Studio compared to Spyder.  It seems Spyder's results are the wrong ones.  It specifically happens when the input array gets large (in mycase, ~1300).  If I shorten the array to 1200, the results are the same, but once past 1300, then the results are different.  I've written about the issues [here](https://stackoverflow.com/questions/74870549/why-am-i-getting-different-results-on-different-platforms-spyder-vs-visual-stud)

For Visual Studio I'm using:
VS: 1.74.2
Python: 3.8.8
Numpy: 1.20.1
Pandas: 1.2.4
sklearn: 0.24.1
scipy: 1.6.2

For Spyder I'm using:
Spyder: 4.2.5
Python: 3.8.8
Numpy: 1.20.1
Pandas: 1.2.4
sklearn: 0.24.1
scipy: 1.6.2

**Comment by StatguyUser** [comment]
What is the version of BaselineRemoval library you are using?

**Comment by rummeln** [comment]
Version 0.1.1

**Comment by StatguyUser** [comment]
Quite odd. Let's debug together. Can you please refresh kernel to clean the memory in Spyder and then run again in Spyder? Can you please also try 2 more methods. Copy paste the code in jupyter and run there? Run your python script file from command line? Please compare the results from jupyter, command line, Spyder and visual studio and tell me which one are correct and where you find discrepancies.

**Comment by StatguyUser** [comment]
Also, your friends results in Spyder matching with your Spyder results? Or are they different?

**Comment by rummeln** [comment]
Okay, let me give it a try.

They're only running it in visual studio, and getting the same results that I am with visual studio.

**Comment by rummeln** [comment]
The results from jupyter and command line match what I see in spyder (which are the incorrect results compared to VS)

**Comment by rummeln** [comment]
I'm trying to confirm the version of baselineremoval in VS but having trouble... saying "BaselineRemoval" has no attribute "version"

**Comment by StatguyUser** [comment]
Did you install BaselineRemoval separately for Spyder and visual studio, using separate commands? Or it was done for once using pip install BaselineRemoval?

Also, what syntax are you using for checking version in visual studio?

**Comment by rummeln** [comment]
ah, ok, so the PS code is actually connected on an HPC in a linux environment.  and my friend is running a mac.  whereas I'm running on a PC.  So it seems the results in from the mac and linux are the same, but with windows I'm getting the different results.  I tried running PScode locally and get the same results as spyder.

I just completely uninstalled and reinstalled everything through the Anaconda powershell prompt.
I'm running:
Python 3.8.10
Scipy 1.5.4
Pandas 1.4.4
Numpy 1.19.5
Scikit-learn 0.23.2
BaselineRemoval 0.1.1

**Comment by rummeln** [comment]
so I think the issue is with my local machine, because we're seeing those baselineremoval ouputs getting "shifted" or "reset" after 1200.  I saw you responded to a similar post about this by saying to upgrade numpy, scipy, and scikitlearn, so I'm doing that now

**Comment by rummeln** [comment]
Getting the same results using:
Numpy 1.24.0
Scipy 1.9.3
Sklearn 1.2.0

any idea why I might be getting different results using windows?

**Comment by rummeln** [comment]
If the array is 1290 the result is fine, but once it's 1291, then the baseline correction seems to falter.  This happens if I use the array [0,0:1291] or [0, 100:1391], so it's not bad column/value.

<img width="1242" alt="1290" src="https://user-images.githubusercontent.com/121206875/209418530-701ffe15-5aec-499c-9397-f60baf6217d7.png">
<img width="1211" alt="1291" src="https://user-images.githubusercontent.com/121206875/209418528-3cab57e6-f77a-403f-866b-c1d153e899af.png">

**Comment by rummeln** [comment]
Also, just noticed that if I use polynomial_degree = 2, the results match (but I've been using polynomial_degree = 3)

**Comment by StatguyUser** [comment]
Ok, so the polynomial degree you have been using is 3 and your friend has been using degree 2? This could be the reason. The default value for polynomials is 2. It might be your friend is using default values and didn't set a different polynomial.

**Comment by StatguyUser** [comment]
Let me know if both of you were using different polynomials or same polynomials.

**Comment by rummeln** [comment]
Sorry for the confusion, we're both using the same polynomials.  When we used 3, we got different results, but when we switch to 2, we get the same.

**Comment by StatguyUser** [comment]
Quite strange. I developed and tested it on Linux originally, where it was working fine. As you have quite a good deal of experience with this issue with windows, would you like to be a co-contributor? If you are ready for this, please make a local copy of the repo and debug. I will help you answer questions if you have any. Once done, we can merge the code and release the new version.

**Comment by rummeln** [comment]
Any idea why it wouldn't work with polynomial 3.  That's the one we'd like to use and unfortunately it has to be run on a windows system.

**Comment by StatguyUser** [comment]
There is nothing in the source code that is sensitive to a different degree of a polynomial or the number of data points. The seed is also set as fixed to allow reproducibility. The only conclusion from this we can draw is that something is happening at the operating system level. Although this is hard to believe. As we can see that everything is the same across both machines, except for the operating system.

**Comment by StatguyUser** [comment]
I hope I am wrong in saying this, but can't think of anything else. Possibly some of the libraries such as scipy, sklearn, or numpy behave differently across the different operating systems. We have used these libraries for some tasks. Or the code and functions I have written behave differently in different operating systems.

**Comment by StatguyUser** [comment]
Please clarify the below points.
1) your results from the visual studio is matching with your friend's results from visual studio on polynomial 3? yes or no?
2) your only issue is matching visual studio results from both your friend and your machine with spyder results from your machine, i.e. windows?

If both are yes, then something to do with spyder IDE. Just trying to find the correct source of error.

**Comment by rummeln** [comment]
Thanks for your input/help, I appreciate the effort you've put in!

1. When I run Visual Studio through my HPC (on linux environment), I get the same results as my friend (who's running on a mac).  This is when using polynomial = 3.
2. The results I get with with Visual Studio run locally (on my local windows environment), I get the same results as what I get when running Spyder.  These results are different from the results from what I got through the HPC (on linux environment) and my friend (on mac).  This is also using polynomial = 3.
3. When I change to polynomial = 2, then the results are the same for all of the systems/environments listed above.

**Comment by StatguyUser** [comment]
So the problem as we concluded is that if we are using a polynomial beyond 2 for modpoly and imodpoly in a windows machine, the results might not be accurate. For other scenarios and operating systems, it works just fine. Thank you for pointing this out.

**Comment by StatguyUser** [comment]
Can you please email me the data? I will like to debug and get to the root of it. I will like to fix the issue. Let me know and I will post my email address.

**Comment by StatguyUser** [comment]
This issue has now been identified and fixed. The source of error was coming from line `np.array(input_array_for_poly)`. In Linux, np.array returns numpy array in numpy.int64 format, whereas for Windows, np.array returns the array in np.int64 format. This is where results were differing. In some cases, int64 was also insufficient. Hence, default data type is now set to `dtype=np.object`

**Comment by StatguyUser** [comment]
This issue has now been identified and fixed. The source of error was coming from line `np.array(input_array_for_poly)`. In Linux, np.array returns numpy array in numpy.int64 format, whereas for Windows, np.array returns the array in np.int32 format. This is where results were differing. In some cases, int64 was also insufficient. Hence, default data type is now set to `dtype=np.object`
